external_executors/dnanexus_wes/dx-wes-lambda/dx_wes.py

```diff
@@ -0,0 +1,510 @@
+    response = table.query(
+        KeyConditionExpression=boto3.dynamodb.conditions.Key('WESWorkflowId').eq(wes_workflow_id) & \
+                               boto3.dynamodb.conditions.Key("EventType").eq(event_type)
+    )
+
+    # If Count is zero, then there's no recorded analysis id so just return
+    # None
+    if not response["Count"]:
+        return None
+
+    value = response["Items"][0][field]
+
+    return value
+
+lookup_dx_analysis_id = functools.partial(_lookup_table, event_type="DxAnalysisId", field="Id")
+lookup_dxwdl_log = functools.partial(_lookup_table, event_type="DxWDLLog", field="Log")
+
+def record_dxwdl_log(wes_workflow_id, log):
+    """Record a log event associated with the execution of dxWDL itself. For
+    the actual workflow execution logs, we can make requests to DNAnexus, but
+    if reports something interesting, we have to record that ourselved in the
+    database.
+    """
+
+    ddb = boto3.resource("dynamodb", region_name=os.environ["AWS_REGION"])
+    table = ddb.Table(os.environ["AWS_DDB_TABLE"])
+    table.put_item(
+        Item={
+            "WESWorkflowId": wes_workflow_id,
+            "EventType": "DxWDLLog",
+            "Log": log
+        }
+    )
+
+def record_dx_analysis_id(wes_workflow_id, dx_analysis_id):
+    """Record an association between a WES workflow id and a DXAnalysis id."""
+
+    ddb = boto3.resource("dynamodb", region_name=os.environ["AWS_REGION"])
+    table = ddb.Table(os.environ["AWS_DDB_TABLE"])
+
+    # If the record already exists, update it. We may have recorded a dummy
+    # value previously.
+    if lookup_dx_analysis_id(wes_workflow_id):
+        table.update_item(
+            Key={
+                "WESWorkflowId": wes_workflow_id,
+                "EventType": "DxAnalysisId"
+            },
+            UpdateExpression="set Id = :p",
+            ExpressionAttributeValues={
+                ":p": dx_analysis_id
+            }
+        )
+    else:
+        table.put_item(
+            Item={
+                "WESWorkflowId": wes_workflow_id,
+                "EventType": "DxAnalysisId",
+                "Id": dx_analysis_id
+            }
+        )
+
+
+def dx_localize_inputs(inputs_dict, dx_project_id, wes_workflow_id):
+    """Localize input files in the inputs dict to DNAnexus.
+
+    The inputs dict may contain URLs that reference remote files, and we
+    need to put those within DNAnexus before creating the workflow so the
+    workflow steps can access the files.
+
+    Also, the inputs dict might be arbitrarily nested, so we have to
+    recursively walk through it to try to find URLS to localize.
+
+    The output of this function should be the a dict with identical structure
+    as the inputs_dict, but with gs and http URLs replaced with dx URLs.
+    """
+
+    def match_url(s):
+        """Return True if s looks like a URL we can do something with."""
+        return s.startswith("http://") or \
```

Comment:
We could probably also pretty easily work with ftp links, I imagine (for the future)

Reply:
Yeah, also `s3` would be pretty easy too I think.